File: 2097-valid-arrangement-of-pairs/2097-valid-arrangement-of-pairs.cpp
class Solution {
public:
    vector<int> path;
    vector<vector<int>> validArrangement(vector<vector<int>>& pairs) {
        unordered_map<int,vector<int>> adj;
        unordered_map<int,int> outDegrees;
        unordered_map<int,int> inDegrees;
        for(int i=0;i<pairs.size();i++) {
            adj[pairs[i][0]].push_back(pairs[i][1]);
            outDegrees[pairs[i][0]]++;
            inDegrees[pairs[i][1]]++;
        }
        int start = pairs[0][0];
        for(auto [node,_]:adj) {
            int diff = outDegrees[node]-inDegrees[node];
            if(diff==1) {
                start = node;
                break;
            }
        }
        dfs(adj,start);
        vector<vector<int>> ans;
        reverse(path.begin(),path.end());
        for(int i=0;i<path.size()-1;i++) {
            ans.push_back({path[i],path[i+1]});
        }
        return ans;
    }

    void dfs(unordered_map<int,vector<int>>& adj,int node) {
        while(!adj[node].empty()) {
            int neigh = adj[node].back();
            adj[node].pop_back();
            dfs(adj,neigh);
        }
        path.push_back(node);
        return;
    }
};



/*
4   5
5   1
9   4
11  9

4 -> 5
5 -> 1
9 -> 4
11 -> 9
*/

File: 2097-valid-arrangement-of-pairs/2097-valid-arrangement-of-pairs.java
import java.util.*;

class Solution {
    private List<Integer> path = new ArrayList<>();

    public int[][] validArrangement(int[][] pairs) {
        Map<Integer, List<Integer>> adj = new HashMap<>();
        Map<Integer, Integer> outDegrees = new HashMap<>();
        Map<Integer, Integer> inDegrees = new HashMap<>();

        // Build adjacency list and degree counts
        for (int[] pair : pairs) {
            adj.computeIfAbsent(pair[0], k -> new ArrayList<>()).add(pair[1]);
            outDegrees.put(pair[0], outDegrees.getOrDefault(pair[0], 0) + 1);
            inDegrees.put(pair[1], inDegrees.getOrDefault(pair[1], 0) + 1);
        }

        // Find start node
        int start = pairs[0][0];
        for (int node : adj.keySet()) {
            int outDegree = outDegrees.getOrDefault(node, 0);
            int inDegree = inDegrees.getOrDefault(node, 0);
            if (outDegree - inDegree == 1) {
                start = node;
                break;
            }
        }

        // Perform DFS to find the Eulerian path
        dfs(adj, start);

        // Construct the result
        Collections.reverse(path);
        int[][] result = new int[path.size() - 1][2];
        for (int i = 0; i < path.size() - 1; i++) {
            result[i][0] = path.get(i);
            result[i][1] = path.get(i + 1);
        }

        return result;
    }

    private void dfs(Map<Integer, List<Integer>> adj, int node) {
        List<Integer> neighbors = adj.getOrDefault(node, new ArrayList<>());
        while (!neighbors.isEmpty()) {
            int next = neighbors.remove(neighbors.size() - 1);
            dfs(adj, next);
        }
        path.add(node);
    }
}
